### PR TITLE
QPFS_filter_method

### PR DESCRIPTION
**Pull request by Top1Miami**
Вот туть фильтр по QPFS

### CODE REVIEW

File: sample.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+from sklearn.datasets import make_classification
```
Comment (LastShekel) [inline sample.py]: What a file is this?

File: ITMO_FS/filters/QPFS.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
+import numpy as np
+from scipy.linalg import sqrtm
+from sklearn.metrics import mutual_info_score
+from qpsolvers import solve_qp
+import math
+
+class QPFS(object):
```
Comment (LastShekel) [inline ITMO_FS/filters/QPFS.py]: As it considered to be filter, this algorithm should be a function not a class or something

**Comment by Top1Miami** [comment]
Requested changes completed

File: ITMO_FS/filters/QPFS.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+def __filterBy(sigma, eigvals, U):
+	if sigma == None:
+		return (eigvals, U)
+	y = []
+	for i in range(len(eigvals)):
+		if eigvals[i] > sigma:
+			y.append(i)
+	return (eigvals[y], U[:, y])
+
+def __count_priors(y):
```
Comment (LastShekel) [inline ITMO_FS/filters/QPFS.py]: sugest you to use numpy.hist

File: ITMO_FS/filters/QPFS.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+	class_size = max(y) + 1
+	priors = np.zeros(class_size)
+	for i in y:
+		priors[i] += 1
+	return list(map(lambda x: x / len(y), priors))
+
+
+def __getCk(y, k):
```
Comment (LastShekel) [inline ITMO_FS/filters/QPFS.py]: numpy where

File: ITMO_FS/filters/QPFS.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+		r = X.shape[1] - 1
+	if r >= X.shape[1]:
+		raise TypeError("r parameter should be less than the number of features")
+	F = np.zeros(X.shape[1]) # F vector represents of each variable with class(here it is intialized)
+	XT = X.T # Transposed matrix X 
+	class_size = max(y) + 1 # Count the number of classes, we assume that class labels would be numbers from 1 to max(y)
+	priors = __count_priors(y) # Count prior probabilities of classes
+	for i in range(class_size): # Loop through classes
+		Ck = __getCk(y, i) # Get array C(i) where C(k) is 1 when i = k and 0 otherwise
+		for j in range(XT.shape[0]): # Loop through features
```
Comment (LastShekel) [inline ITMO_FS/filters/QPFS.py]: redo it with matrix operations

File: ITMO_FS/filters/QPFS.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+			F[j] = priors[i] * fn(XT[j], Ck) # Counting F vector
+	Q = np.zeros((X.shape[1], X.shape[1])) # Q represents dependency between variables(here it is initalized)
+	for i in range(XT.shape[0]): # Loop through features
+		for j in range(XT.shape[0]): # Loop through features
+			Q[i][j] = fn(XT[i], XT[j]) # Counting dependency, using normalized mutual info score
```
Comment (LastShekel) [inline ITMO_FS/filters/QPFS.py]: fn(XT,XT) doesnt work?

File: ITMO_FS/filters/QPFS.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+        The name of qp solver according to qpsolvers(https://pypi.org/project/qpsolvers/) naming.
+        Note quadprog is used by default.
+    fn : function(array, array), default
+        The function to count correlation, for example pierson correlation or  mutual information.
+        Note mutual information is used by default.
+    Returns
+    ------
+    array-like, shape (n_features) : the ranks of features in dataset, with rank increase, feature relevance increases and redundancy decreases.
+    """
+
```
Comment (LastShekel) [inline ITMO_FS/filters/QPFS.py]: Add usage example here

File: ITMO_FS/filters/QPFS.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+	x = np.array([[1, 2, 3, 2, 2], [2, 3, 1, 2, 3], [1, 3, 5, 1, 1], [3, 1, 4, 3, 1], [3, 1, 2, 3, 1]])
+	y = np.array([1, 1, 2, 1, 2])
+	ranks = QPFS().run(x, y)
+	print(ranks)
+"""
+
+def QPFS_filter(X, y, r = None, sigma = None, solv='quadprog', fn=mutual_info_score):#TODO find the r and sigma values to be set as default
+	"""
+    Runs the QPFS algorithm on the specified dataset.
+    A different loss function is used here than in the paper due to using (-1, 1) labels.
```
Comment (LastShekel) [inline ITMO_FS/filters/QPFS.py]: Add paper link

File: ITMO_FS/filters/QPFS.py (hunk 4 above)
Comment (Top1Miami) [reply ITMO_FS/filters/QPFS.py]: what for?

File: ITMO_FS/filters/QPFS.py (hunk 6 above)
Comment (Top1Miami) [reply ITMO_FS/filters/QPFS.py]: Either I have to pass two functions into method or leave it as it is, cause if u take a look at F matrix counting, u will see that fn(XT, XT) wouldnt work anyway.

File: ITMO_FS/filters/QPFS.py (hunk 3 above)
Comment (Top1Miami) [reply ITMO_FS/filters/QPFS.py]: what for?

**Comment by Top1Miami** [comment]
Requested changes done

**Comment by Top1Miami** [comment]
Update!!!!

File: ITMO_FS/filters/QPFS.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+	y = []
+	for i in range(len(eigvals)):
+		if eigvals[i] > sigma:
+			y.append(i)
+	return (eigvals[y], U[:, y])
+
+def __countAlpha(A, B, F):
+	sumQ = 0
+	Comb = B.T.dot(np.linalg.pinv(A)).dot(B)
+	for i in range(A.shape[0]):
```
Comment (LastShekel) [inline ITMO_FS/filters/QPFS.py]: np.sum

File: ITMO_FS/filters/QPFS.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+		for j in range(A.shape[1]):
+			sumQ += A[i][j]
+	for i in range(B.shape[0]):
+		for j in range(B.shape[1]):
```
Comment (LastShekel) [inline ITMO_FS/filters/QPFS.py]: 2*np.sum

File: ITMO_FS/filters/QPFS.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+			sumQ += 2 * B[i][j]
+	for i in range(Comb.shape[0]):
+		for j in range(Comb.shape[1]):
```
Comment (LastShekel) [inline ITMO_FS/filters/QPFS.py]: np.sum

**Comment by Top1Miami** [comment]
DONE!

**Comment by Top1Miami** [comment]
Split wrapper filter

File: ITMO_FS/filters/qpfs_filter.py [hunk 12]
```diff
@@ -0,0 +1,43 @@
+from ITMO_FS.utils.qpfs_body import qpfs_body
+from ITMO_FS.filters.measures import pearson_corr
+
+def qpfs_filter(X, y, r = None, sigma = None, solv='quadprog', fn=pearson_corr):
```
Comment (LastShekel) [inline ITMO_FS/filters/qpfs_filter.py]: move to measures

**Comment by Top1Miami** [comment]
НУ ПАЖАЛУЙСТА

**Pull request merged**